notation.go

```diff
@@ -63,31 +44,189 @@ type SignOptions struct {
+	// descriptor upon successful verification.
+	// If nil signature is present and the verification level is not 'skip',
+	// an error will be returned.
+	Verify(ctx context.Context, signature []byte, opts VerifyOptions) (ocispec.Descriptor, *VerificationOutcome, error)
+}
+
+// Verify performs signature verification on each of the notation supported
+// verification types (like integrity, authenticity, etc.) and return the
+// verification outcomes.
+// For more details on signature verification, see
+// https://github.com/notaryproject/notaryproject/blob/main/specs/trust-store-trust-policy.md#signature-verification
+func Verify(ctx context.Context, verifier Verifier, repo registry.Repository, opts VerifyOptions) (ocispec.Descriptor, []*VerificationOutcome, error) {
+	var verificationOutcomes []*VerificationOutcome
+	artifactRef := opts.ArtifactReference
+	// passing nil signature to check 'skip'
+	_, outcome, err := verifier.Verify(ctx, nil, opts)
+	if err != nil {
+		if outcome == nil {
+			return ocispec.Descriptor{}, nil, err
+		}
+	} else if outcome.VerificationLevel.Name == trustpolicy.LevelSkip.Name {
+		return ocispec.Descriptor{}, []*VerificationOutcome{outcome}, nil
+	}
+
+	// get signature manifests
+	var targetArtifactDesc ocispec.Descriptor
+	artifactDescriptor, err := repo.Resolve(ctx, artifactRef)
+	if err != nil {
+		return ocispec.Descriptor{}, nil, ErrorSignatureRetrievalFailed{Msg: err.Error()}
+	}
+	count := 0
+	err = repo.ListSignatures(ctx, artifactDescriptor, func(signatureManifests []ocispec.Descriptor) error {
+		if count >= verificationSignatureNumLimit {
+			return fmt.Errorf("number of signatures attempted to be verified should be less than: %d", verificationSignatureNumLimit)
+		}
+		// process signatures
+		for _, sigManifestDesc := range signatureManifests {
+			// get signature envelope
+			sigBlob, _, err := repo.FetchSignatureBlob(ctx, sigManifestDesc)
+			if err != nil {
+				return ErrorSignatureRetrievalFailed{Msg: fmt.Sprintf("unable to retrieve digital signature with digest %q associated with %q from the registry, error : %v", sigManifestDesc.Digest, artifactRef, err.Error())}
+			}
+			payloadArtifactDescriptor, outcome, err := verifier.Verify(ctx, sigBlob, opts)
+			if err != nil {
+				if outcome == nil {
+					// TODO: log fatal error
+					return err
+				}
+				verificationOutcomes = append(verificationOutcomes, outcome)
+				continue
+			}
+			if !content.Equal(payloadArtifactDescriptor, artifactDescriptor) {
+				outcome.Error = errors.New("content descriptor mismatch")
+				verificationOutcomes = append(verificationOutcomes, outcome)
+				continue
+			}
+
+			// At this point, we've found a signature verified successfully
+			outcome.RawSignature = sigBlob
+			verificationOutcomes = append(verificationOutcomes, outcome)
+			// Descriptor of the signature blob that gets verified successfully
+			targetArtifactDesc = payloadArtifactDescriptor
+
+			return errDoneVerification
```

Comment:
Are we returning from verification on the first successful signature here? We wanted to process all signatures so that `notation inspect` command can show details of each signature. Similar to `signtool.exe` showing all signatures rather than just until a good signature.

Reply:
IMO, when we have one signature verified successfully, the whole verification is succeeded. In this case, general users would most likely only interest in the result or the succeeded signature. 
We will process all the signatures if none of them gets verified successfully. In this case, users might interest more in the failure details of the signatures.
Adding another option into VerifyOption seems a little bit over designed? It's making the system more complicated. @shizhMSFT any comments/suggestions/concerns on this one?

Reply:
@shizhMSFT This is not out of scope for this PR. Before this PR, we are processing all signatures. The reason for processing all signatures is to support commands like notation inspect. With this refactoring, we are changing the functionality. I'm okay with that, but let's expose this as an option so that library consumers have a way.

Reply:
Let's add an option so that invokers can choose to verify all signatures although we have a valid one.

Reply:
@rgnote By going through the logic, I find there is an issue on the requirement of processing all signatures.

What if we have 100 signatures where the 42th one valid and others are invalid? Since it is over 50 signatures, the `Verify()` function will return error. Then the invoker will think the entire verification process failed although it in fact succeeded.

Reply:
Opened https://github.com/notaryproject/notation-go/issues/198 to track it.